servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/DefaultRequestTracker.java

```diff
@@ -93,15 +93,10 @@ public void onSuccess(final long startTimeNanos) {
     }
 
     @Override
-    public void onCancel(final long startTimeNanos) {
+    public void onError(final long startTimeNanos, ErrorClass errorClass, Throwable cause) {
         pendingUpdater.decrementAndGet(this);
-        calculateAndStore(this::cancelPenalty, startTimeNanos);
-    }
-
-    @Override
-    public void onError(final long startTimeNanos) {
-        pendingUpdater.decrementAndGet(this);
-        calculateAndStore(this::errorPenalty, startTimeNanos);
+        calculateAndStore(errorClass == ErrorClass.CANCELLED ? this:: cancelPenalty : this::errorPenalty,
+                startTimeNanos);
```

Comment:
Looking at how it's implemented here and taking RS spec into account, I would recommend keeping `onCancel` separate for now. It's a special beast and it's not always clear how to deal with it. The biggest challenge is that cancel can race with completion signals, both `onSuccess` and `onError`. For example, users can still get the valid response immediately after they cancel. Another one is that depending on what operators were used, you can see observe cancel and any terminal (`.whenCancel(...).whenOnSuccess(...).whenOnError(...)`) event or only one of 3 (`.whenFinally(TerminalSignalConsumer)`). Then these points became 2 times worse, when you have to deal with `Single<MetaData>` and `Publisher<Buffer>` independently.

Reply:
In my minds eye this isn't a type that needs to be tied to the RS types and it's up to the RS bridge to manage those nuances. I'd argue that keeping them separate methods encourages the call sights to think the `RequestTracker` will "do the right thing" whereas the current implementation can potentially eat two completions: one for a cancel and another for the success/error as you pointed out. This would present itself as a bug since the pending request counter would be decremented twice for the same request.

Reply:
If you'd like to consolidate, consider adding javadoc, describing that `onError` can be invoked multiple times (for cancel and real error) and that it's possible to get `onSuccess` before/after `onError(CANCELLED)` to make sure whoever implements `RequestTracker` takes this into account.

Reply:
I think we need to go the other way: users of `RequestTracker` are on the hook for doing the right thing and not calling multiple `onX` methods for the same `beforeStart()` call: there is no way for `RequestTracker` to know that multiple calls to the `onX` methods correlated to a single call to `beforeStart()` and that will mess up the pending request counter.

Reply:
Then it's worth clarifying these assumptions in javadoc for users of `RequestTracker` instead of implementers

Reply:
I added a header to the `RequestTracker` to more clearly describe the contract.